pkg/dependency/parser/nodejs/pnpm/parse.go

```diff
@@ -127,55 +252,102 @@ func (p *Parser) parseLockfileVersion(lockFile LockFile) float64 {
 	}
 }
 
-// cf. https://github.com/pnpm/pnpm/blob/ce61f8d3c29eee46cee38d56ced45aea8a439a53/packages/dependency-path/src/index.ts#L112-L163
-func (p *Parser) parsePackage(depPath string, lockFileVersion float64) (string, string) {
-	// The version separator is different between v5 and v6+.
-	versionSep := "@"
-	if lockFileVersion < 6 {
-		versionSep = "/"
+func (p *Parser) parseDepPath(depPath string, lockVer float64) (string, string) {
+	dPath := p.trimRegistry(depPath, lockVer)
+
+	var scope string
+	scope, dPath = p.separateScope(dPath)
+
+	var name string
+	name, dPath = p.separateName(dPath, lockVer)
+
+	// add scope to pkg name
+	if scope != "" {
+		name = fmt.Sprintf("%s/%s", scope, name)
 	}
-	return p.parseDepPath(depPath, versionSep)
+
+	ver := p.trimPeerDeps(dPath, lockVer)
+
+	return name, ver
 }
 
-func (p *Parser) parseDepPath(depPath, versionSep string) (string, string) {
-	// Skip registry
-	// e.g.
-	//    - "registry.npmjs.org/lodash/4.17.10" => "lodash/4.17.10"
-	//    - "registry.npmjs.org/@babel/generator/7.21.9" => "@babel/generator/7.21.9"
-	//    - "/lodash/4.17.10" => "lodash/4.17.10"
-	_, depPath, _ = strings.Cut(depPath, "/")
+// trimRegistry trims registry (or `/` prefix) for depPath.
```

Comment:
PURL has `vcs_url`. I think we should keep this information if it is not `registry.npmjs.org`.
https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm

Reply:
updated in https://github.com/aquasecurity/trivy/pull/6617/commits/1dc202a7556f11d0835aa56bec43d11a84954a2a

Reply:
I am not sure if using registry URL in package name is correct.
I suggest keeping depPath in [ExternalReferences](https://github.com/aquasecurity/trivy/blob/bbaf5952bc8059c537401bd4364e019eecd16c9a/pkg/fanal/types/package.go#L173)
and check `ExternalReferences` in purl for npm:
https://github.com/aquasecurity/trivy/blob/bbaf5952bc8059c537401bd4364e019eecd16c9a/pkg/purl/purl.go#L425-L429

e.g.
`private.npmjs.org/@babel/runtime@7.18.3` =>
```Go
ID:           "@babel/runtime@7.18.3",
Name:         "@babel/runtime",
Version:      "7.18.3",
Relationship: ftypes.RelationshipIndirect,
ExternalReferences: []ftypes.ExternalRef{
	{
		Type: ftypes.RefVCS,
		URL:  "private.npmjs.org/@babel/runtime@7.18.3",
	},
},
```

@knqyf263 wdyt?

Reply:
Yes, I meant to suggest using `ExternalReferences`, lack of explanation.

Reply:
I update this logic - https://github.com/aquasecurity/trivy/pull/6617/commits/2d50522c0d18104c8cb5b4c628ed3b75cc326352
take a look, when you have time, please